src/ProducerEventProcessor/ProducerEventProcessorTest.cpp

```diff
@@ -1321,33 +1326,207 @@ TEST(ProducerEventProcessor, ThreadNameSmoke) {
+                          ::testing::IsTrue()),
+      ::testing::Property("thread_state_slice", &ClientCaptureEvent::thread_state_slice,
+                          ThreadStateSliceEq(expected_thread_state_slice)));
+}
+
+::testing::Matcher<ClientCaptureEvent> ClientCaptureEventsInternedCallstackEq(
+    const InternedCallstack& expected_interned_callstack) {
+  return ::testing::AllOf(
+      ::testing::Property("has_interned_callstack", &ClientCaptureEvent::has_interned_callstack,
+                          ::testing::IsTrue()),
+      ::testing::Property("interned_callstack", &ClientCaptureEvent::interned_callstack,
+                          InternedCallstackEq(expected_interned_callstack)));
+}
+
 TEST(ProducerEventProcessor, ThreadStateSliceSmoke) {
   MockClientCaptureEventCollector collector;
   auto producer_event_processor = ProducerEventProcessor::Create(&collector);
 
   ProducerCaptureEvent producer_event;
-  {
-    ThreadStateSlice* thread_state_slice = producer_event.mutable_thread_state_slice();
-    thread_state_slice->set_pid(kPid1);
-    thread_state_slice->set_tid(kTid1);
-    thread_state_slice->set_thread_state(ThreadStateSlice::kIdle);
-    thread_state_slice->set_duration_ns(kDurationNs1);
-    thread_state_slice->set_end_timestamp_ns(kTimestampNs1);
-  }
+  ThreadStateSlice* producer_thread_state_slice = producer_event.mutable_thread_state_slice();
+  producer_thread_state_slice->set_pid(kPid1);
+  producer_thread_state_slice->set_tid(kTid1);
+  producer_thread_state_slice->set_thread_state(ThreadStateSlice::kIdle);
+  producer_thread_state_slice->set_duration_ns(kDurationNs1);
+  producer_thread_state_slice->set_end_timestamp_ns(kTimestampNs1);
+  producer_thread_state_slice->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kNoCallstack);
+  producer_thread_state_slice->set_switch_out_or_wakeup_callstack_id(0);
+
+  ThreadStateSlice expected_thread_state_slice = *producer_thread_state_slice;
 
   ClientCaptureEvent event;
-
   EXPECT_CALL(collector, AddEvent).Times(1).WillOnce(SaveArg<0>(&event));
 
-  producer_event_processor->ProcessEvent(1, std::move(producer_event));
+  producer_event_processor->ProcessEvent(orbit_grpc_protos::kLinuxTracingProducerId,
+                                         std::move(producer_event));
 
   ASSERT_EQ(event.event_case(), ClientCaptureEvent::kThreadStateSlice);
-  const ThreadStateSlice& thread_state_slice = event.thread_state_slice();
-  EXPECT_EQ(thread_state_slice.pid(), kPid1);
-  EXPECT_EQ(thread_state_slice.tid(), kTid1);
-  EXPECT_EQ(thread_state_slice.thread_state(), ThreadStateSlice::kIdle);
-  EXPECT_EQ(thread_state_slice.duration_ns(), kDurationNs1);
-  EXPECT_EQ(thread_state_slice.end_timestamp_ns(), kTimestampNs1);
+  EXPECT_THAT(event.thread_state_slice(), ThreadStateSliceEq(expected_thread_state_slice));
+}
+
+TEST(ProducerEventProcessor, ThreadStateSliceMergesCallstack) {
```

Comment:
The test name sounds a bit strange. I propose "MergingThreadStateSliceWithCallstack".
And then the one below can be called "MergingThreadStateSliceWithCallstackFailsGracefullyWhenNoCallstackWasSeen", which is now a proper sentence.

Reply:
Done.